nil/internal/collate/validator.go

```diff
@@ -2,26 +2,91 @@ package collate
 
 import (
 	"context"
+	"encoding/json"
+	"errors"
 	"fmt"
+	"sync"
 
 	"github.com/NilFoundation/nil/nil/common"
+	"github.com/NilFoundation/nil/nil/common/check"
+	"github.com/NilFoundation/nil/nil/common/logging"
+	"github.com/NilFoundation/nil/nil/internal/config"
 	"github.com/NilFoundation/nil/nil/internal/db"
 	"github.com/NilFoundation/nil/nil/internal/execution"
 	"github.com/NilFoundation/nil/nil/internal/network"
+	"github.com/NilFoundation/nil/nil/internal/signer"
 	"github.com/NilFoundation/nil/nil/internal/types"
 	"github.com/rs/zerolog"
 )
 
+var (
+	errOldBlock         = errors.New("received old block")
+	errOutOfOrder       = errors.New("received block is out of order")
+	errHashMismatch     = errors.New("block hash mismatch")
+	errInvalidSignature = errors.New("invalid block signature")
+)
+
 type Validator struct {
-	params Params
+	params *Params
 
 	txFabric       db.DB
 	pool           TxnPool
 	networkManager *network.Manager
+	blockVerifier  *signer.BlockVerifier
+	lastBlock      *types.Block
+	lastBlockHash  common.Hash
 
+	mutex  sync.RWMutex
 	logger zerolog.Logger
 }
 
+func NewValidator(params *Params, txFabric db.DB, pool TxnPool, nm *network.Manager) *Validator {
+	return &Validator{
+		params:         params,
+		txFabric:       txFabric,
+		pool:           pool,
+		networkManager: nm,
+		blockVerifier:  signer.NewBlockVerifier(params.ShardId, txFabric),
+		logger: logging.NewLogger("validator").With().
+			Stringer(logging.FieldShardId, params.ShardId).
+			Logger(),
+	}
+}
+
+func (s *Validator) getLastBlockUnlocked(ctx context.Context) (*types.Block, common.Hash, error) {
+	if s.lastBlock != nil {
+		return s.lastBlock, s.lastBlockHash, nil
+	}
+
+	rotx, err := s.txFabric.CreateRoTx(ctx)
+	if err != nil {
+		return nil, common.EmptyHash, err
+	}
+	defer rotx.Rollback()
+
+	block, hash, err := db.ReadLastBlock(rotx, s.params.ShardId)
+	if err != nil && !errors.Is(err, db.ErrKeyNotFound) {
+		return nil, common.EmptyHash, err
+	}
+	if err == nil {
+		return block, hash, err
+	}
+	return nil, common.EmptyHash, nil
+}
+
+func (s *Validator) GetLastBlock(ctx context.Context) (*types.Block, common.Hash, error) {
```

Comment:
I'm not completely sure that putting fine-grained locks into individual functions is a good idea. Can we do something like a lock around all operations with the validator? It's very easy to miss funny edge cases otherwise.

Reply:
I'll think about it. Probably all public functions should have lock management and call only one `<functionName>Unlocked` function inside.

You are basically right but I'm a bit confused by choosing correct approach here.